### PR TITLE
Start implementation of multivariate posterior in mcmcstep

### PR DESCRIPTION
**Pull request by miaoqingyu2**
I've started adding some function to calculate the posterior distribution. to align with previous discussion, i plan to use the inverse formula of $\Sigma$ and $\Sigma_\mu$. Specifically, 

leaf node follows $\mathcal{N}( (n_l \Sigma^{-1} + \Sigma_\mu^{-1})^{-1} \Sigma^{-1} \sum r_i, (n_l \Sigma^{-1} + \Sigma_\mu^{-1})^{-1} )$

and likelihood ratio becomes
\[
\text{likelihood ratio} \;=\;
\sqrt{
\frac{
\det\!\Big(\tfrac{n_\ell}{n_1 n_2}\Sigma_\mu^{-1}\Big)\;
\det\!\big(\Sigma^{-1}+\tfrac{1}{n_\ell}\Sigma_\mu^{-1}\big)
}{
\det\!\big(\Sigma^{-1}+\tfrac{1}{n_1}\Sigma_\mu^{-1}\big)\;
\det\!\big(\Sigma^{-1}+\tfrac{1}{n_2}\Sigma_\mu^{-1}\big)
}
}
\;\times\;
\exp\!\Bigg(
\tfrac{1}{2}\,\tilde{R}_1^{\!\top}\Sigma^{-1}\big(\Sigma^{-1}+\tfrac{1}{n_1}\Sigma_\mu^{-1}\big)^{-1}\Sigma^{-1}\tilde{R}_1
+ \tfrac{1}{2}\,\tilde{R}_2^{\!\top}\Sigma^{-1}\big(\Sigma^{-1}+\tfrac{1}{n_2}\Sigma_\mu^{-1}\big)^{-1}\Sigma^{-1}\tilde{R}_2
- \tfrac{1}{2}\,\tilde{R}_\ell^{\!\top}\Sigma^{-1}\big(\Sigma^{-1}+\tfrac{1}{n_\ell}\Sigma_\mu^{-1}\big)^{-1}\Sigma^{-1}\tilde{R}_\ell
\Bigg).
\]

### CODE REVIEW

**Comment by miaoqingyu2** [comment]
> I left some comments.
> 
> So I guess I got the expression for the posterior wrong, I placed `n_l` on the other matrix! Your version looks like the right one because it kind of divides the sum-of-residuals by the number of points in the leaf, which makes sense to get an average.
> 
> Please also double-check my expression for the likelihood ratio if you haven't.

The likelihood ratio I calculated is:
$$
\sqrt{
\frac{ \det(\Sigma_{\mu}^{-1}) \det( \Sigma_{\mu}^{-1} + n_\ell \Sigma^{-1} )  }
{ \det( \Sigma_{\mu}^{-1} + n_{\ell l} \Sigma^{-1} )   \det( \Sigma_{\mu}^{-1} + n_{\ell r} \Sigma^{-1} )  }
}
\exp{
\frac{1}{2} \tilde{R}_{\ell l}^T \Sigma^{-1} ( \Sigma_{\mu}^{-1} + n_{\ell l} \Sigma^{-1})^{-1} \Sigma^{-1} \tilde{R}_{\ell l}
+ \frac{1}{2} \tilde{R}_{\ell r}^T \Sigma^{-1} ( \Sigma_{\mu}^{-1} + n_{\ell r} \Sigma^{-1})^{-1} \Sigma^{-1} \tilde{R}_{\ell r}
- \frac{1}{2} \tilde{R}_{\ell}^T \Sigma^{-1} ( \Sigma_{\mu}^{-1} + n_{\ell} \Sigma^{-1})^{-1} \Sigma^{-1} \tilde{R}_{\ell}
}
$$

File: src/bartz/mcmcstep.py [hunk 1]
```diff
@@ -2073,6 +2129,49 @@ def precompute_leaf_terms(
...
+    Returns
+    -------
+    Pre-computed terms for leaf sampling in multivariate case.
+    """
+    num_trees, num_leaves = prec_trees.shape
+    k = sigma2_cov_prec.shape[0]
+
+    n_k = prec_trees[..., None, None]  # Shape: [num_trees, num_leaves, 1, 1]
+    posterior_precision = sigma_mu2_cov_inv + n_k * sigma2_cov_prec
+    L_prec = jnp.linalg.cholesky(posterior_precision)
```
[the remark this answers is not in the document]
Comment (miaoqingyu2) [reply src/bartz/mcmcstep.py:2162]: I think it should be fine here, because the precision and prior matrices are assumed to be positive definite as a prerequisite. But we could add a warning or check before the Cholesky step if we want to be extra safe.

[the remark this answers is not in the document]
Comment (Gattocrucco) [reply src/bartz/mcmcstep.py]: Positive definite != well conditioned

[the remark this answers is not in the document]
Comment (Gattocrucco) [reply src/bartz/mcmcstep.py]: As the number of dimensions increases, it's quite common to work with badly conditioned covariance matrices.

**Review by Gattocrucco — CHANGES_REQUESTED** [review]
I only checked the old stuff (cholesky & inverses), left a comment. I still haven't checked the rest of the new code from the last commit because it'll take a while, I wanted to at least comment immediately on the stuff I could check quickly.

File: src/bartz/mcmcstep.py [hunk 2]
```diff
@@ -2073,6 +2161,59 @@ def precompute_leaf_terms(
...
+    """
+    num_trees, num_leaves = prec_trees.shape
+    k = error_cov_inv.shape[0]
+    n_k = prec_trees[..., None, None]  # Shape: [num_trees, num_leaves, 1, 1]
+    error_cov_inv_batched = jnp.broadcast_to(
+        error_cov_inv, (num_trees, num_leaves, k, k)
+    )
+    leaf_prior_cov_inv_batched = jnp.broadcast_to(
+        leaf_prior_cov_inv, (num_trees, num_leaves, k, k)
+    )
```
[the remark this answers is not in the document]
Comment (Gattocrucco) [reply src/bartz/mcmcstep.py:2200]: I checked the documentation of `solve_triangular`, it looks like it requires explicit shapes indeed.

Remove `leaf_prior_cov_inv_batched` though, that does look redundant.

[the remark this answers is not in the document]
Comment (Gattocrucco) [reply src/bartz/mcmcstep.py:2200]: Also add a comment that `error_cov_inv_batched` is for satisfying the batching rules of `solve_triangular`.

File: src/bartz/mcmcstep.py [hunk 3]
```diff
@@ -2073,6 +2129,49 @@ def precompute_leaf_terms(
...
+    k = sigma2_cov_prec.shape[0]
+
+    n_k = prec_trees[..., None, None]  # Shape: [num_trees, num_leaves, 1, 1]
+    posterior_precision = sigma_mu2_cov_inv + n_k * sigma2_cov_prec
+    L_prec = jnp.linalg.cholesky(posterior_precision)
+    Y = solve_triangular(L_prec, sigma2_cov_prec, lower=True)
+    mean_factor = solve_triangular(L_prec, Y, lower=True, trans='T')
+
+    z = random.normal(key, (num_trees, num_leaves, k))
+    centered_leaves = solve_triangular(L_prec.T, z, lower=False)
```
[the remark this answers is not in the document]
Comment (Gattocrucco) [reply src/bartz/mcmcstep.py]: => I read `solve_triangular`'s docstring and `z` does not need the additional axis.
